### ISSUE TITLE
Reverse planner broken, need help understand a code snippet

### ISSUE THREAD

**Issue by Sebastianv650** [MarlinFirmware/Marlin#9913]
I think I found the error causing issues #9728 and #9494. Basicaly, every move starts and ends at MINIMAL_STEP_RATE of 120.

This is due to the fact that the reverse planner exits it's loop if:
```cpp
if (TEST(current->flag, BLOCK_BIT_START_FROM_FULL_HALT)) // Up to this every block is already optimized.
  break;
```

This bit is set during the junction jerk calculation if the planner was empty. But there is also a second case which enables this bit, which I don't understand:
```cpp
const float vmax_junction_threshold = vmax_junction * 0.99f;
if (previous_safe_speed > vmax_junction_threshold && safe_speed > vmax_junction_threshold) {
  // Not coasting. The machine will stop and start the movements anyway,
  // better to start the segment from start.
  SBI(block->flag, BLOCK_BIT_START_FROM_FULL_HALT);
  vmax_junction = safe_speed;
}
```

The comment isn't very helpful. What's the meaning of this check, @thinkyhead maybe you can bring some light into this as you did the implementation of Prusas jerk code?
The jerk code which is setting this bit worked once before, because before #9149 the entry speed was set to the max. possible so no reverse planner was needed. But this behaviour was the reason for #9093 (exceeded jerk values when buffer runs dry).

I think the solution should be easy when the reason for this second bit-set-case is known.

**Comment by Sebastianv650** [comment]
Maybe this is true:
They are separating three different states:
- Coasting, which means each axis continues the movement in the same direction
- Reversal of an axis
- Starting or stopping an axis.

This if statement covers the second and third case. Removing the SBI line "solves" the problem as the reverse planner will th**e**n adjust the speed as needed.
But even then there is the problem that the last block of a series is always starting and ending at MINIMAL_STEP_RATE. This is due to the fact that the newest block is never reverse planned.

**Comment by thinkyhead** [comment]
> What's the meaning of this check, @thinkyhead maybe you can bring some light into this as you did the implementation of Prusas jerk code?

See my (probably useless) answer [here](https://github.com/MarlinFirmware/Marlin/pull/9914#issuecomment-370180970).

**Comment by thinkyhead** [comment]
> This is due to the fact that the newest block is never reverse planned.

Right! What a weird thing. That's what I was trying to track down and fix when I posted #9360, but I didn't solve it at that time.

Let's have a peek at MK3 firmware and see what they've done with this in the interim…

`BLOCK_FLAG_START_FROM_FULL_HALT` is still there, hmm.

https://github.com/prusa3d/Prusa-Firmware/blob/MK3/Firmware/planner.cpp#L344
```cpp
        // No need to recalculate the last block, it has already been set by the plan_buffer_line() function.
        // Vojtech thinks, that one shall not touch the entry speed of the very first block as well, because
        // 1) it may already be running at the stepper interrupt,
        // 2) there is no way to limit it when going in the forward direction.
        while (block_index != tail) {
            if (current->flag & BLOCK_FLAG_START_FROM_FULL_HALT) {
              . . .
```

**Comment by Sebastianv650** [comment]
> // No need to recalculate the last block, it has already been set by the plan_buffer_line() function.

That's the mistake which creates extremely jerky movements as soon as the buffer drains. Prusas changes looks as they have tried to make the reverse planner somewhat obsolete. They tried to set the speed inside plan_buffer_line, which can only work when the buffer is always well filled, therefore they skipped the reverse planner in most cases.
